### DISCUSSION TITLE
SMV Script - Headless Rendering for QGIS2FDS verification

### DISCUSSION THREAD

**Discussion by johodges** [firemodels/fds#12016] [General]
I have a first cut at Github actions for running qgis2fds verification cases on my fork. Unfortunately, because QGIS packages different versions of external libraries with the same QGIS version on different OS, there are small differences in the binary bingeom files generated which makes it difficult to add a rule for acceptance at the end of the action. Rather than looking for an exact match in the binary data, I was hoping to use the same algorithm firebot uses to check the delta between images generated from SMV running the generated FDS case. Unfortunately, SMV hangs when I try to generate the images with a headless display. Has anyone else used SMV in this way?

**Comment by JakeOShannessy** [comment]
I don't believe the way Smokeview works currently that you'll be able to run it headless. Have you looked into virtual framebuffers? That was a solution we were looking into but never implemented.

**Reply by johodges** [reply]
Thanks for the suggestion. I'll read up on the topic and see if it will work for this application.

**Reply by JakeOShannessy** [reply]
**NB**: Xvfb (as per Glenn's comment below) is a virtual framebuffer, so sounds like that's the way to go.

**Comment by gforney** [comment]
both firebot and smokebot generate images using Xvfb on our linux cluster - our head node has a monitor but these scripts do not use it to generate the images

**Comment by gforney** [comment]
look at fds/Verification/scripts/Make_FDS_Pictures.sh in the fds repo

**Reply by johodges** [reply]
Thanks, will take a look.

**Comment by gforney** [comment]
you will also need to install Mesa (opengl without a video card), and some X libraries.

**Comment by johodges** [comment]
Thanks for pointing me to the verification scripts. It was straightforward to implement in the github action after reviewing the files. I did not have to add any additional packages to get Xvfb running in the instance since QGIS had many of the same dependencies. For future reference these are the relevant commands:

- Find the display port "id -u" (1001 in the case for this machine used in this configuration)
- Set DISPLAY environmental variable  - DISPLAY: ":1001"
- Run Xvfb - "Xvfb :1001 -fp /usr/share/X11/fonts/misc -screen 0 1280x1024x24 &"
- Pause some number of seconds to allow Xvfb to spin up "sleep 5"
- Run smokeview with the -runscript option

**Comment by gforney** [comment]
great!

the script firebot and smokebot uses to compare images is located at bot/Firebot/compare_images.sh in the bot repo

search for the string 'convert' - that is where the comparison is done.  I blur the two images slightly to be compared first.  It is an attempt to minimizes false positives (a positive  meaning two images are different).  The command convert is from the imagemagick package

**Reply by johodges** [reply]
Thanks, I was able to integrate a slightly modified version of compare_images.sh for use in the qgis2fds github actions. I should have the ubuntu automated suite ready this week.
